### PR TITLE
TASK-29: Login Backend

### LINKED ISSUE #111: TASK-29: Login Backend
**Task Description**

As a developer, I want the backend for logging in so that I work on the frontend logic.

**Acceptance Criteria**

- [ ] Have JWT Tokens
  - [ ] Access tokens present
  - [ ] Refresh tokens present
- [ ] Have a GraphQL mutation for logging in
- [ ] Have a REST route for refresh tokens
  - [ ] The route is reached via a POST request
  - [ ] A cookie is set on the client containing the refresh token
  - [ ] The cookie is secure on HTTPS and the cookie is only sent on this route 
- [ ] Documentation present where necessary
- [ ] Unit and UI tests are implemented for every component and function

### PR DESCRIPTION
**Pull request by bit172**
This PR closes #111.

**Steps to Test Local**

1. Start the server using ```nx serve mull-api```
2. Open the GraphQL playground.
3. Click the Settings cog on the top right
4. Change the ```request.credentials``` setting to "include"
 Like so: 
image
5. Open your browser dev tools in the ```Application``` tab
6. Under Storage, click on cookies for the server domain (http://localhost:3333)
7. Create a user using a mutation or mull-ui
8. Login using the ```login``` mutation
9. An accessToken should be returned and a cookie named ```mullToken``` should be set.

**Steps to Test OAuth**
1. Start the server using ```nx serve mull-api```
2. Open your browser and go to ```http://localhost:3333/api/auth/google | twitter | facebook/```
3. An accessToken should be returned and a cookie named ```mullToken``` should be set.

**Notes**
Refer to the sequence diagrams on [lucidchart](https://lucid.app/lucidchart/invitations/accept/6afcb749-5988-4bdb-9def-865d6dcf9ad9) for more information.
This requires an update to the .env file. Updated .env available on [Slack](https://bug-bytes.slack.com/archives/CS5P4NL5Q/p1606083803001700).

### CODE REVIEW

**Review by ys-lin — DISMISSED** [review]
This looks great! :) 
  - [x] Have JWT Tokens
  - [x] Access tokens present
  - [x] Refresh tokens present
- [x] Have a GraphQL mutation for logging in
- [x] Have a REST route for refresh tokens
  - [x] The route is reached via a POST request
  - [x] A cookie is set on the client containing the refresh token
  - [x] The cookie is secure on HTTPS and the cookie is only sent on this route 
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function

File: apps/mull-api/src/app/auth/auth.service.ts [hunk 1]
```diff
@@ -1,49 +1,67 @@
...
+    return this.jwtService.sign(
+      { id: user.id, tokenVersion: user.tokenVersion },
+      { expiresIn: '7d', secret: environment.jwt.refreshSecret }
+    );
+  }
+
+  sendRefreshToken(res: Response, refreshToken: string) {
+    res.cookie('mullToken', refreshToken, {
+      maxAge: 7 * 24 * 3600 * 1000,
+      secure: environment.production,
```
Comment (r-saba) [inline apps/mull-api/src/app/auth/auth.service.ts]: A comment might be helpful so future readers don't have to go through documentation 
```suggestion
      secure: environment.production, // determines if cookie is to be used with HTTPS only 
```

File: apps/mull-api/src/app/app.module.ts [hunk 2]
```diff
@@ -49,6 +49,11 @@ import { AuthModule } from './auth/auth.module';
       uploads: {
         maxFileSize: 10000000000000,
       },
+      cors: {
+        origin: environment.client.baseUrl,
```
Comment (r-saba) [inline apps/mull-api/src/app/app.module.ts]: Think we can add [Apollo Studio](https://studio.apollographql.com/) during dev only? We would add https://studio.apollographql.com/ to the `origin` property here and in `main.ts`

**Review by r-saba — DISMISSED** [review]
Most of my time was spent understanding the flow and implementation, didn't find any issues LGTM!

Since you'll be rebasing anyway, there's 2 small comments I left that you could add in if you'd like.

- [X] Have JWT Tokens
  - [X] Access tokens present
  - [X] Refresh tokens present
- [X] Have a GraphQL mutation for logging in
- [x] Have a REST route for refresh tokens
  - [X] The route is reached via a POST request
  - [X] A cookie is set on the client containing the refresh token
  - [X] The cookie is secure on HTTPS and the cookie is only sent on this route 
- [X] Documentation present where necessary
- [X] Unit and UI tests are implemented for every component and function

**Review by r-saba — APPROVED** [review]
My approval from above still stands. Also thanks for adding Apollo Studio support in dev

**Review by ys-lin — APPROVED** [review]

**Pull request merged**